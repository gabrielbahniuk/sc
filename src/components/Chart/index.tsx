import React, { useRef, useEffect } from 'react'
import { Chart as ChartJs } from 'chart.js'
import useTimeSeries from '@/hooks/useTimeSeries'
import { appConfig } from '@/config'

type ConeProps = {
  riskLevel: number
  mu: number
  sigma: number
}

type ChartProps = {
  yearsInvestment: number
  cones: ConeProps[]
  investmentValue: number
  riskLevel: number
}

const Chart: React.FC<ChartProps> = ({ investmentValue, yearsInvestment, riskLevel, cones }) => {
  const canvas = useRef<HTMLCanvasElement>(null)

  const { mu, sigma } = cones.filter((cone) => cone.riskLevel === riskLevel)[0]
  const { FEE, MONTHLY_SUM } = appConfig
  ChartJs.defaults.global.defaultFontColor = '#fff'

  const { chartData } = useTimeSeries({
    years: yearsInvestment,
    initialSum: investmentValue,
    fee: FEE,
    mu,
    sigma,
    monthlySum: MONTHLY_SUM
  })

  useEffect(() => {
    drawChart()
  }, [])

  function drawChart(): void {
    const { labels, dataBad, dataGood, dataMedian } = chartData
    const data = {
      datasets: [
        {
          data: dataGood,
          label: 'Good performance',
          borderColor: 'rgba(100, 255, 100, 0.4)',
          fill: false,
          pointRadius: 1
        },
        {
          data: dataMedian,
          label: 'Median performance',
          borderColor: 'rgba(100, 100, 100, 0.4)',
          fill: false,
          pointRadius: 1
        },
        {
          data: dataBad,
          label: 'Bad performance',
          borderColor: 'rgba(255, 100, 100, 0.4)',
          fill: false,
          pointRadius: 1
        }
      ],
      labels
    }

    const options = {
      responsive: false,
      scales: {
        xAxes: [
          {
            display: true,
            scaleLabel: {
              display: true,
              labelString: 'Years',
              fontSize: 16
            },
            gridLines: {
              drawOnChartArea: false
            }
          }
        ],
        yAxes: [
          {
            display: true,
            scaleLabel: {
              display: true,
              fontSize: 16,
              labelString: 'Valuation (EUR)'
            }
          }
        ]
      }
    }

    const config = {
      type: 'line',
      data,
      options
    }

    const context = canvas.current?.getContext('2d')
    new ChartJs(context, config)
  }

  return <canvas ref={canvas} />
}

export default Chart
