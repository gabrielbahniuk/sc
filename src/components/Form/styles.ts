import styled from 'styled-components'

export const FormContainer = styled.form`
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 80%;

  label {
    font-size: 1.5rem;
    color: var(--text-color);
    width: 100%;
  }
  input,
  select {
    width: 100%;
    padding: 0 1.5rem;
    height: 4rem;
    border-radius: 0.25rem;
    border: 1px solid #d7d7d7;
    background: #e7e9ee;
    font-weight: 400;
    font-size: 1.5rem;
  }

  input[type='number']::-webkit-inner-spin-button,
  input[type='number']::-webkit-outer-spin-button {
    -webkit-appearance: none;
    -webkit-appearance: textfield;
    -moz-appearance: textfield;
    appearance: textfield;

    cursor: pointer;
    display: block;
    width: 1rem;
    color: #333;
    text-align: center;
    position: relative;
  }

  label + label {
    display: block;
    margin-top: 2rem;
  }
`
