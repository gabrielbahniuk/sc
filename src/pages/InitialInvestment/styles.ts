import styled from 'styled-components'

export const Container = styled.div`
  max-width: 1120px;
  margin: 1rem auto;
  padding: 2.5rem 1rem;
  display: flex;
  flex-direction: column;
  align-items: center;

  h1 {
    text-align: center;
    color: var(--text-color);
    margin-bottom: 3rem;
    font-size: 2.5rem;
  }
`
