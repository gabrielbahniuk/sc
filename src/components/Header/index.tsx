/* eslint-disable @typescript-eslint/explicit-function-return-type */
import React from 'react'
import Logo from '@/assets/logo.svg'

import { Container } from './styles'

const Header: React.FC = () => {
  return (
    <Container>
      <Logo />
    </Container>
  )
}
export default Header
