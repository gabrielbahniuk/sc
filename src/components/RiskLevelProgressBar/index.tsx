import React from 'react'

import { ProgressBar, ProgressLabelContainer } from './styles'

type RiskLevelProgressBar = {
  riskLevel: number
}

const RiskLevelProgressBar: React.FC<RiskLevelProgressBar> = ({ riskLevel }) => {
  return (
    <>
      <ProgressBar riskLevel={riskLevel}>
        <span data-cy="progressBar"></span>
      </ProgressBar>
      <ProgressLabelContainer>
        <span>Low Risk</span>
        <span>High Risk</span>
      </ProgressLabelContainer>
    </>
  )
}

export default RiskLevelProgressBar
