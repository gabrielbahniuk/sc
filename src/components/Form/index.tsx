import React from 'react'
import { useHistory } from 'react-router-dom'
import { useInvestments } from '@/hooks'
import RiskLevelSelector from '@/components/RiskLevelSelector'
import RiskLevelProgressBar from '@/components/RiskLevelProgressBar'
import Button from '@/components/Button'

const Form: React.FC = () => {
  const history = useHistory()
  const {
    initialInvestment,
    riskLevel,
    amountInvestmentYears,
    handleInitialInvestment,
    handleRiskLevel,
    handleAmountInvestmentYears,
    MINIMUM_INITIAL_INVESTMENT,
    MINIMUM_YEARS_INVESTMENT,
    MAXIMUM_YEARS_INVESTMENT
  } = useInvestments()

  function handleNextPage(): void {
    history.push('/result')
  }

  return (
    <form>
      <label>
        Initial Investment
        <input
          type="number"
          step="1000"
          data-cy="initialInvestment"
          min={MINIMUM_INITIAL_INVESTMENT}
          value={initialInvestment}
          onChange={handleInitialInvestment}
        />
      </label>
      <label>
        How many years?
        <input
          type="number"
          min={MINIMUM_YEARS_INVESTMENT}
          max={MAXIMUM_YEARS_INVESTMENT}
          step="1"
          data-cy="amountInvestmentYears"
          value={amountInvestmentYears}
          onChange={handleAmountInvestmentYears}
        />
      </label>
      <label>
        Risk Level
        <RiskLevelSelector data-cy="riskLevelSelector" riskLevel={riskLevel} handleRiskLevel={handleRiskLevel} />
      </label>
      <RiskLevelProgressBar riskLevel={riskLevel} />
      <Button data-cy="btnCalculate" onClick={handleNextPage}>
        Calculate
      </Button>
    </form>
  )
}

export default Form
