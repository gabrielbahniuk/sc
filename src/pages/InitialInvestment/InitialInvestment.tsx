import React from 'react'
import { Form } from '@/components/Form'

import { Container } from './styles'

const InitialInvestment: React.FC = () => {
  return (
    <>
      <Container>
        <h1>Investments Calculator</h1>
        <Form />
      </Container>
    </>
  )
}

export { InitialInvestment }
