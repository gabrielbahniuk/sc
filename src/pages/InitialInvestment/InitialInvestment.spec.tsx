import React from 'react'
import { shallow, ShallowWrapper } from 'enzyme'
import { InitialInvestment } from './InitialInvestment'
import { Form } from '@/components/Form'
import { create } from 'react-test-renderer'

describe('InitialInvestment Page', () => {
  it('should match snapshot', () => {
    const initialInvestmentPage = create(<InitialInvestment />).toJSON()
    expect(initialInvestmentPage).toMatchSnapshot()
  })
  it('should render page components', () => {
    const wrapper: ShallowWrapper = shallow(<InitialInvestment />)
    expect(wrapper.find('h1').length).toEqual(1)
    expect(wrapper.find(Form).length).toEqual(1)
  })
})
