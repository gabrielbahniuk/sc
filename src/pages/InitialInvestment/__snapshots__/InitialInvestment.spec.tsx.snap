// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InitialInvestment Page should match snapshot 1`] = `
<div
  className="sc-eCApnc kXjkOl"
>
  <h1>
    Investments Calculator
  </h1>
  <form
    className="sc-hKFxyN dsJelp"
  >
    <label>
      Initial Investment
      <input
        data-cy="initialInvestment"
        step="1000"
        type="number"
      />
    </label>
    <label>
      How many years?
      <input
        data-cy="amountInvestmentYears"
        step="1"
        type="number"
      />
    </label>
    <label>
      Risk Level
      <select
        data-cy="riskLevelSelector"
      >
        <option
          value={3}
        >
          3
        </option>
        <option
          value={4}
        >
          4
        </option>
        <option
          value={5}
        >
          5
        </option>
        <option
          value={6}
        >
          6
        </option>
        <option
          value={7}
        >
          7
        </option>
        <option
          value={8}
        >
          8
        </option>
        <option
          value={9}
        >
          9
        </option>
        <option
          value={10}
        >
          10
        </option>
        <option
          value={11}
        >
          11
        </option>
        <option
          value={12}
        >
          12
        </option>
        <option
          value={13}
        >
          13
        </option>
        <option
          value={14}
        >
          14
        </option>
        <option
          value={15}
        >
          15
        </option>
        <option
          value={16}
        >
          16
        </option>
        <option
          value={17}
        >
          17
        </option>
        <option
          value={18}
        >
          18
        </option>
        <option
          value={19}
        >
          19
        </option>
        <option
          value={20}
        >
          20
        </option>
        <option
          value={21}
        >
          21
        </option>
        <option
          value={22}
        >
          22
        </option>
        <option
          value={23}
        >
          23
        </option>
        <option
          value={24}
        >
          24
        </option>
        <option
          value={25}
        >
          25
        </option>
      </select>
    </label>
    <div
      className="sc-bdnxRM fUQRin"
    >
      <span
        data-cy="progressBar"
      />
    </div>
    <div
      className="sc-gtsrHT htXvXo"
    >
      <span>
        Low Risk
      </span>
      <span>
        High Risk
      </span>
    </div>
    <button
      className="sc-dlnjwi gafhhG"
      data-cy="btnCalculate"
      onClick={[Function]}
    >
      Calculate
    </button>
  </form>
</div>
`;
